Doc/library/asyncio-sync.rst

```diff
@@ -285,11 +259,54 @@ Semaphore
 
 
 BoundedSemaphore
-^^^^^^^^^^^^^^^^
+----------------
 
 .. class:: BoundedSemaphore(value=1, \*, loop=None)
 
    A bounded semaphore implementation. Inherit from :class:`Semaphore`.
 
    This raises :exc:`ValueError` in :meth:`~Semaphore.release` if it would
    increase the value above the initial value.
+
+   Locks also support the :ref:`context management protocol <async-with-locks>`.
+
+   This class is :ref:`not thread safe <asyncio-multithreading>`.
+
+
+.. _async-with-locks:
+
+Using locks, conditions, and semaphores in the :keyword:`async with` statement
+------------------------------------------------------------------------------
+
+All of the objects provided by this module that have :meth:`acquire`
+and :meth:`release` methods can be used as context managers for a
+:keyword:`async with` statement.  The :meth:`acquire` method will be
+called when the block is entered, and :meth:`release` will be called
+when the block is exited.  Hence, the following snippet::
+
+   async with lock:
+       # do something...
+
+is equivalent to::
+
+   await lock.acquire()
+   try:
+       # do something...
+   finally:
+       lock.release()
+
+Currently, :class:`Lock`, :class:`Condition`,
+:class:`Semaphore`, and :class:`BoundedSemaphore` objects may be used as
+:keyword:`async with` statement context managers.
+
+``async with lock`` is the preferable way for locking asyncio
+synchronization primitives.
+
+Explicit :meth:`acquire` / :meth:`release` calls should be used if
+locking/unlocking is split into different functions.
```

Comment:
I don't think we need these two paragraphs.  People usually naturally prefer 'with' blocks when they are available.